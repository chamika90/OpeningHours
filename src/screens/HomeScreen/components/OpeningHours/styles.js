import {StyleSheet} from 'react-native';
import {theme} from '../../../../config/theme';

const {colors} = theme;

export const styles = StyleSheet.create({
  openHoursContainer: {
    borderRadius: 10,
    backgroundColor: colors.primaryCardBackgroundColor,
    shadowColor: colors.primaryShadowColor,
    shadowOffset: {width: 0, height: 2},
    shadowOpacity: 1,
    shadowRadius: 2,
    padding: 30,
    margin: 10,
    elevation: 10,
  },
  header: {
    alignItems: 'flex-start',
    marginBottom: 10,
    flexDirection: 'column',
  },
  titleContainer: {
    flexDirection: 'row',
  },
  title: {
    color: colors.primaryTextColor,
    fontSize: 24,
    fontFamily: 'Roboto-Bold',
  },
  headerIcon: {
    marginRight: 15,
    justifyContent: 'center',
  },
  line: {
    width: '100%',
    height: 1,
    marginTop: 12,
    backgroundColor: colors.primaryTextColor,
  },
  listSeparator: {
    width: '100%',
    height: 1,
    backgroundColor: colors.primaryListSeparatorColor,
  },
  rowContainer: {
    flexDirection: 'row',
    paddingVertical: 10,
  },
  nameContainer: {
    flex: 2,
    flexDirection: 'row',
    alignItems: 'center',
  },
  day: {
    fontFamily: 'Roboto-Medium',
    fontSize: 12,
    color: colors.primaryTextColor,
    textAlign: 'center',
  },
  timeDurationText: {
    fontFamily: 'Roboto-Regular',
    fontSize: 12,
    color: colors.primaryTextColor,
  },
  closeText: {
    color: colors.secondaryTextColor,
    fontFamily: 'Roboto-Regular',
    fontSize: 12,
  },
  today: {
    fontFamily: 'Roboto-Bold',
    fontSize: 10,
    color: colors.textGreenColor,
    marginLeft: 6,
    textAlign: 'center',
  },
  hoursContainer: {
    flex: 3,
    flexDirection: 'row',
    alignItems: 'flex-end',
    justifyContent: 'flex-end',
  },
});
