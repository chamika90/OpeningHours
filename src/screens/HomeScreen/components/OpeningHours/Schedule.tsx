import React, {useMemo} from 'react';
import {FlatList, Text, View} from 'react-native';
import {days} from '../../../../config/constants';
import {
  convertFirstLetterToUpperCase,
  getCurrentDay,
} from '../../../../helper/utils';
import {styles} from './styles';

export type ScheduleProps = {
  [dayOfWeek: string]: string;
};
type ScheduleRowProps = {
  day: string;
};
export type HoursListProps = {
  hoursList: Array<OpeningHour>;
};

export type OpeningHour = {
  type: string;
  value: number;
};

type OpenHoursListProps = {
  hoursList: string;
};

const Schedule = ({schedule}: {schedule: ScheduleProps}) => {
  const currentDay = useMemo(() => getCurrentDay(), []);

  const OpenHoursList = ({hoursList}: OpenHoursListProps) => {
    return (
      <View>
        {(hoursList !== '' && (
          <Text
            style={styles.timeDurationText}
            adjustsFontSizeToFit
            testID="hoursListText">
            {hoursList}
          </Text>
        )) || (
          <Text testID="closeText" style={styles.closeText}>
            Closed
          </Text>
        )}
      </View>
    );
  };

  const ScheduleRow = ({day}: ScheduleRowProps) => {
    const formatedDay = convertFirstLetterToUpperCase(day);
    const isToday = day === currentDay.toLowerCase();
    return (
      <View testID="scheduleRow" style={styles.rowContainer}>
        <View style={styles.nameContainer}>
          <Text style={styles.day}>{formatedDay}</Text>
          {isToday && <Text style={styles.today}>TODAY</Text>}
        </View>

        {schedule && (
          <View testID="openHoursList" style={styles.hoursContainer}>
            <OpenHoursList hoursList={schedule[day] ?? ''} />
          </View>
        )}
      </View>
    );
  };

  const ListSeporator = () => {
    return <View style={styles.listSeparator} />;
  };

  return (
    <FlatList
      testID="schedule"
      data={days}
      ItemSeparatorComponent={() => <ListSeporator />}
      renderItem={({item}) => <ScheduleRow day={item} />}
      ListFooterComponent={() => <ListSeporator />}
    />
  );
};

export default Schedule;
