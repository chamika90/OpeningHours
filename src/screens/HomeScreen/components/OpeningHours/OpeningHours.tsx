import React from 'react';
import {Text, View} from 'react-native';
import {useRestaurantOpenHours} from '../../../../hooks/useRestaurantOpenHours';
import Icon from 'react-native-vector-icons/FontAwesome5';
import {theme} from '../../../../config/theme';
import Loader from '../../../../components/Loader/Loader';
import Schedule from './Schedule';
import {styles} from './styles';
import ErrorComponent from '../../../../components/ErrorComponent/ErrorComponent';

const {colors} = theme;

const OpeningHours = () => {
  const {openHours, isDataLoading, error} = useRestaurantOpenHours();

  type TitleProps = {
    title: string;
  };
  const Title = ({title}: TitleProps) => {
    return (
      <View style={styles.header}>
        <View style={styles.titleContainer}>
          <View style={styles.headerIcon}>
            <Icon name="clock" size={18} color={colors.primaryIconColor} />
          </View>
          <Text testID="title" style={styles.title}>
            {title}
          </Text>
        </View>
        <View style={styles.line} />
      </View>
    );
  };

  return (
    <View style={styles.openHoursContainer}>
      <Title title={'Opening hours'} />
      {openHours && <Schedule schedule={openHours} />}
      <Loader isLoading={isDataLoading} />
      {error && (
        <ErrorComponent message="Something went wrong. please try again later" />
      )}
    </View>
  );
};

export default OpeningHours;
