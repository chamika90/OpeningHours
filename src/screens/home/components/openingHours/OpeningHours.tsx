import React from 'react';
import {View} from 'react-native';
import {useRestaurantOpenHours} from '../../../../hooks/useRestaurantOpenHours';
import Title from './components/Title';
import Loader from '../../../../components/loader/Loader';
import Schedule from './components/Schedule';
import {styles} from './styles';
import ErrorComponent from '../../../../components/errorComponent/ErrorComponent';

const OpeningHours = () => {
  const {openHours, isDataLoading, error} = useRestaurantOpenHours();

  return (
    <View style={styles.openHoursContainer}>
      <Title title={'Opening hours'} />
      {openHours && <Schedule schedule={openHours} />}
      <Loader isLoading={isDataLoading} />
      {error && (
        <ErrorComponent message="Something went wrong. please try again later" />
      )}
    </View>
  );
};

export default OpeningHours;
