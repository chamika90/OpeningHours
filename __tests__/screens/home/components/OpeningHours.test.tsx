import 'react-native';
import React from 'react';
import {render, waitFor} from '@testing-library/react-native';
require('jest-fetch-mock').enableMocks();
import OpeningHours from '../../../../src/screens/HomeScreen/components/OpeningHours/OpeningHours';

describe('Opening Hours ', () => {
  it('should match the opening hours snapshot', () => {
    const {toJSON} = render(<OpeningHours />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render the title ', () => {
    const {getByTestId} = render(<OpeningHours />);
    expect(getByTestId('title')).toBeTruthy();
  });

  it('should be initially loading', async () => {
    const {getByTestId} = render(<OpeningHours />);

    await waitFor(() => {
      expect(getByTestId('loader')).toBeDefined();
    });
  });
});
