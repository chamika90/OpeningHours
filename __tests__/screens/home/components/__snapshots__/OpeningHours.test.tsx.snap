// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Opening Hours  should match the opening hours snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 10,
      "margin": 10,
      "padding": 30,
      "shadowColor": "#A1A2A4",
      "shadowOffset": Object {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 1,
      "shadowRadius": 2,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "column",
        "marginBottom": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "justifyContent": "center",
            "marginRight": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#A1A2A4",
                "fontSize": 18,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome5Free-Regular",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {
                "fontWeight": "400",
              },
            ]
          }
        >
          ï€—
        </Text>
      </View>
      <Text
        style={
          Object {
            "color": "#202125",
            "fontFamily": "Roboto-Bold",
            "fontSize": 24,
          }
        }
        testID="title"
      >
        Opening hours
      </Text>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#202125",
          "height": 1,
          "marginTop": 10,
          "width": "100%",
        }
      }
    />
  </View>
  <ActivityIndicator
    animating={true}
    color="#A1A2A4"
    size="large"
    style={
      Object {
        "left": "50%",
        "position": "absolute",
        "top": "50%",
      }
    }
    testID="loader"
  />
</View>
`;
